src/main/java/org/jabref/logic/openoffice/StyleLoader.java

```diff
@@ -53,28 +57,54 @@ public StyleLoader(OpenOfficePreferences openOfficePreferences, LayoutFormatterP
      * Adds the given style to the list of styles
      *
      * @param filename The filename of the style
-     * @return True if the added style is valid, false otherwise
+     * @return parse log. result.hasError() is false if the style is added, true otherwise.
+     *         // was: True if the added style is valid, false otherwise
      */
-    public boolean addStyleIfValid(String filename) {
+    public OOBibStyleParser.ParseLog addStyleIfValid(String filename) {
         Objects.requireNonNull(filename);
         try {
-            OOBibStyle newStyle = new OOBibStyle(new File(filename), layoutFormatterPreferences, encoding);
+            OOBibStyle newStyle = new OOBibStyle(new File(filename),
+                                                 layoutFormatterPreferences,
+                                                 encoding);
+
+            OOBibStyleParser.ParseLog parseLog = newStyle.getParseLog();
+            if ( parseLog == null ) {
```

Comment:
It seems that the IDE used was not configured according to https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace. Thus, checkstyle will fail.

Reply:
Does each commit have to pass checkstyle?
I hoped it is enough to check from time to time, and before PRs.

Reply:
Each commit doesn't have to pass Checkstyle, but it becomes easier to review if they pass or are close to passing Checkstyle. Is there anything except the line-wrapping issue that discourages you from using the default settings?

Reply:
> Is there anything except the line-wrapping issue that discourages
> you from using the default settings?

I assume "default settings" above refers to "checkstyle with JabRef
settings". I am using those now, through a call to checkstyle with JabRef
settings in my Makefile before "gradlew build". Earlier I was relying
on visits to Intellij IDEA to check from time to time, and by regularly
postponing this I collected many non-confirming spaces.

> the line-wrapping issue

This appeared in two situations.

One is when I formatted code in order to read it.
Now I removed many (all?) of those.

I understand it is not a good idea to mix formatting with actual changes,
and often separated these into "format" and "fit-to-width" commits.
But these were still interspersed with actual changes.

This is why I asked about "fit-to-width" PRs: these would separate
code formatting and code changes. So far my impression is that
these are considered unnecessary among the people here,
probably because -- as you have shown above -- it is less of a problem
for Intellij IDEA users. Unfortunately I am not ready to switch to it.


The other situation is with new or changed code:
```
class X {
    public static  //x
    ALongDescriptiveNameOfATemplate<ALongParameter, AndOfcourseAnAnotherOne> //x
    descriptiveMethodName( //x
        LongParameterType firstParameter,
        Another secondParameter) {
   }
}
```

my impression is that none of the line breaks marked with `//x` are
considered OK (even if checkstyle accepts some of them). Taken together,
they can result in long lines, like this:
```
class X {
    public static ALongDescriptiveNameOfATemplate<ALongParameter, AndOfcourseAnAnotherOne> descriptiveMethodName(LongParameterType firstParameter,
                                                                                                                 Another secondParameter) {
   }
}
```



**On the question of independent PRs**: although I could point to some
changes that could be considered for thematic splitting, these are small when
compared to a large body of restructuring. (I removed changes to the *.jstyle parser
as well as changes outside openoffice directories (except JabRef_en.properties)).

GUI changes: there are no changes to the OO Panel or other parts of
the normally visible GUI.

Changes to gui-related code consist so far basically of

- (g1) replacing series of calls to OOBibBase by a single call (moving
    detailed knowledge of OOBibBase working from GUI stuff to an
    OOBibBase method dedicated to the given GUI function)

- (g2) catching more exceptions (with associated display of warning
    dialogs)

- (g3) checking more preconditions (with associated display of warning
    dialogs).

OO interface changes: What is changed as seen by the user in Libreoffice?

- (o1) unresolved citations are not thrown away, they get an
  "Unresolved(citationKey)" representation in the text (unless using
  numbered citations) and they also get a bibliography entry with a
  similar content

- (o2) these bibliography entries now contain clickable
  cross-references pointing to the corresponding citation groups in
  the text.

- (o3) GUI actions now wrapped into OO Undo groups, so their effects
  can be undone with a single click (except those affecting pageInfo in
  custom properties)

- (o4) parts of GUI actions are wrapped into "disable screen
  refresh" which reduces flashing. The exception is the part
  deciding "visual order", which relies on asking the position of the
  text view cursor.

- (o5) more careful sorting of footnote markers now allows Merge to
  more reliably discover consecutive citation groups in footnotes.

- (o6) citation groups with no intervening space can now be merged

- (o7) Even with Settings/Automatically sync.. off, leave recognizable marks

    Problem was:
    - Turn Settings/Automatically sync.. off on OO/LO panel
    - Select an article, Click Cite
    - The reference mark created was empty.

    It turned out, that in OOBibBase.insertReferenceMark
    position.setString(citText); failed to set the text.

    Solution was: create a new cursor and use that.


OOBibStyle changes:
- (j1) (Removed parser changes. Note: the parser does need changes:
        silently ignoring user input the parser does not understand, providing
        no warning on mistyped keys, like for example
        BibtexKeyCitations (as appears in the official exemples:
        jabref/src/main/resources/resource/openoffice/default_authoryear.jstyle,
        jstyles.jabref.org-master/turabian-english.jstyle,
        jstyles.jabref.org-master/turabian-deutsch.jstyle
        and jstyles.jabref.org-master/remoteSensingOfEnvironment.jstyle)
        instead of the expected BibTeXKeyCitations,
        inheriting settings from a previous parse, ignoring changes if
        they only affect the LAYOUT section make it a nightmare
        figuring out why some changes to the style have no effect)

- (j2) xxxMarkupBefore xxxMarkupAfter

    *.jstyle now allows insertion of markup before and after a citation group
             as well as some of its parts

    implemented
       CitationGroupMarkupBefore   and CitationGroupMarkupAfter
       AuthorsPartMarkupBefore     and AuthorsPartMarkupAfter
       AuthorNamesListMarkupBefore and AuthorNamesListMarkupAfter
       AuthorNameMarkupBefore      and AuthorNameMarkupAfter

   This was an opportunistic extension, allowed by "Most text we emit
   into the document handled as OOFormattedText". The latter was needed
   for proper handling of italicization of "et al.", that is doing it
   (using markup) when creating the citation marker, not heuristically
   by search-and-edit EtAlString when inserting.

Of these
- (g1) and (g2) are part of the restructuring.
- (g3) and (o3) were implemented to help during testing-by-hand
- (o4) was not necessary for testing. It just follows a similar pattern as
  the Undo part and I happened to find the corresponding API in the OO docs.
- (o5) In principle could have been left out. The hypothesis about
  what could be the problem (why citation groups in footnotes are not
  merged) of course came from the effort to understand what does the
  code do. I could have left it as a hypothesis with some notes on
  what do I think, and why, and which parts of the code should be
  changed to attempt a solution. And after the extra effort to create
  these notes, then let's say a month later, I could scratch my head what was I
  thinking and where are those parts now. As time goes, I do forget things,
  so it is time to locate and reread the corresponding code ... Doing things
  the administratively appropriate way, while it has its virtues, 
  may result in losing the opportunity to do them at al.

Reply:
> I assume "default settings" above refers to "checkstyle with JabRef settings".

Yup, sorry about the confusion.

> So far my impression is that these are considered unnecessary among the people here,

You are right.

> Taken together, they can result in long lines, like this:

Unfortunately, I'd suggest going with the long lines even if they aren't always enforced by Checkstyle.

> This is why I asked about "fit-to-width" PRs: these would separate code formatting and code changes.

In practice, when reviewing a PR I'd be using the [Files changed](https://github.com/JabRef/jabref/pull/7469/files) tab in the PR and it easily becomes "cluttered" if I happen to look at a commit that is "fit-to-width".

> On the question of independent PRs: although I could point to some changes that could be considered for thematic splitting, these are small when compared to a large body of restructuring.

Can they be split without the larger body of restructuring? Even (g1), (g2), (g3), and (o3) seem like a large PR in itself but it would allow us to better understand what you are changing and why in a more incremental way. Frankly, even (g2) by itself could be considered a good contribution.

Reply:
> In practice, when reviewing a PR I'd be using the Files changed tab
> in the PR and it easily becomes "cluttered" if I happen to look at a commit that is "fit-to-width".
Yes, it is clutter against actual changes. The idea was to separate them into distinct PRs that
only contain formatting changes.

> Can they be split without the larger body of restructuring?
Before, after, or in the middle of restructuring?
Some of them can probably be removed and added again (after).
Like not catching an exception. Or in other cases to 
also omit the checks that throw an early warning.
But my impression is that on the way from macaroni to
separation of concerns alot of code moves, some of them repeatedly. 

Currently I am trying to separate decisions about formatting (job for the style)
and inserting text according to this format. Much of this latter is done
in insertOOFormattedTextAtCurrentLocation, but some of the formatting decisions
are interspersed with  inserting and formatting  in populateBibTextSection.
To make the separation possible, I am extending the "language" understood
by insertOOFormattedTextAtCurrentLocation (for example with a <p> tag)

Reply:
Thank you, it is appreciated!

This is with the intent of branching out parts of `OOUtil.java` and `OOBibBase.java` into its own PR or specific commits?

I'll start reading up on those files and hopefully, I'll be able to come with some more concrete suggestions afterward.

Reply:
The last version you can see here is old. Since that I started a new branch, rebased, cut back, then continued.
Should I push the current branch?

Reply:
The summary from upstream/main to HEAD below suggests
that
- (1) the suggested max 400 lines per PR would correspond to
10456/400 = 26 or more PRs. Calculating with 1 PR per week, that is half year.
I hope we can do with much less PRs.
- (2) a strategy for the presentation might be to build up the alternative
version without integrating it, then with the infrastructure in place
switch to it. A consequence of this strategy is that until we get to
the integration part, the new code is not used, you cannot test
it by compiling and running a JabRef. But it is probably easier on the eye.

```
56 files changed, 10456 insertions(+), 2092 deletions(-)
.../gui/openoffice/BibEntryNotFoundException.java  |   16 -
.../gui/openoffice/CitationEntryViewModel.java     |    2 +-
.../gui/openoffice/ManageCitationsDialogView.java  |    3 +-
.../openoffice/ManageCitationsDialogViewModel.java |   46 +-
.../java/org/jabref/gui/openoffice/OOBibBase.java  | 2914 ++++++++++++--------
.../org/jabref/gui/openoffice/OpenOfficePanel.java |  152 +-
.../gui/openoffice/StyleSelectDialogView.java      |    4 +-
.../gui/openoffice/StyleSelectDialogViewModel.java |    4 +-
.../gui/openoffice/StyleSelectItemViewModel.java   |    2 +-
.../java/org/jabref/logic/layout/LayoutEntry.java  |    2 +-
.../java/org/jabref/logic/oostyle/Citation.java    |   54 +
.../logic/oostyle/CitationDatabaseLookup.java      |   31 +
.../{openoffice => oostyle}/CitationEntry.java     |    2 +-
.../org/jabref/logic/oostyle/CitationGroup.java    |  102 +
.../org/jabref/logic/oostyle/CitationGroupID.java  |   18 +
.../org/jabref/logic/oostyle/CitationGroups.java   |  323 +++
.../jabref/logic/oostyle/CitationMarkerEntry.java  |   61 +
.../logic/oostyle/CitationMarkerEntryImpl.java     |  108 +
.../org/jabref/logic/oostyle/CitationPath.java     |   15 +
.../org/jabref/logic/oostyle/CitationSort.java     |   60 +
.../java/org/jabref/logic/oostyle/CitedKey.java    |   81 +
.../java/org/jabref/logic/oostyle/CitedKeys.java   |   75 +
src/main/java/org/jabref/logic/oostyle/Compat.java |  156 ++
.../logic/{openoffice => oostyle}/OOBibStyle.java  |  590 +++-
.../logic/oostyle/OOBibStyleGetCitationMarker.java |  748 +++++
.../oostyle/OOBibStyleGetNumCitationMarker.java    |  386 +++
.../java/org/jabref/logic/oostyle/OOFormat.java    |   80 +
.../org/jabref/logic/oostyle/OOFormattedText.java  |   28 +
.../{openoffice => oostyle}/OOPreFormatter.java    |    2 +-
.../java/org/jabref/logic/oostyle/OOProcess.java   |  473 ++++
.../logic/{openoffice => oostyle}/StyleLoader.java |    3 +-
.../org/jabref/logic/openoffice/Backend52.java     |  383 +++
.../java/org/jabref/logic/openoffice/Codec52.java  |  113 +
.../openoffice/CreationException.java              |    4 +-
.../logic/openoffice/DocumentConnection.java       | 1310 +++++++++
.../openoffice/NoDocumentException.java            |    4 +-
.../org/jabref/logic/openoffice/OOFrontend.java    |  633 +++++
.../java/org/jabref/logic/openoffice/OOUtil.java   |  829 ++++--
.../logic/openoffice/RangeForOverlapCheck.java     |   39 +
.../org/jabref/logic/openoffice/RangeKeyedMap.java |   77 +
.../jabref/logic/openoffice/RangeKeyedMapList.java |  121 +
.../org/jabref/logic/openoffice/RangeSort.java     |   54 +
.../jabref/logic/openoffice/RangeSortVisual.java   |  260 ++
.../org/jabref/logic/openoffice/StorageBase.java   |   92 +
.../logic/openoffice/StorageBaseRefMark.java       |  532 ++++
.../logic/openoffice/UndefinedBibtexEntry.java     |   21 -
.../UndefinedCharacterFormatException.java         |    4 +-
.../org/jabref/preferences/JabRefPreferences.java  |    2 +-
src/main/resources/l10n/JabRef_en.properties       |   57 +-
.../{openoffice => oostyle}/CitationEntryTest.java |    2 +-
.../org/jabref/logic/oostyle/OOBibStyleTest.java   |  949 +++++++
.../OOPreFormatterTest.java                        |    2 +-
.../{openoffice => oostyle}/StyleLoaderTest.java   |    3 +-
.../jabref/logic/openoffice/OOBibStyleTest.java    |  516 ----
.../logic/{openoffice => oostyle}/test.jstyle      |    0 
.../testWithDefaultAtFirstLIne.jstyle              |    0 
```

Reply:
This is not a PR, just a copy of the current state of my version 
https://github.com/antalk2/jabref/tree/improve-reversibility-rebased-03

Reply:
> The last version you can see here is old. Since that I started a new branch, rebased, cut back, then continued.
> Should I push the current branch?

1. You rebased on main?
   2. Is this your "point A"?
2. With "cut back" you mean that you changed back the code style?

---

On the topic of what @calixtus was saying before,
> "Fixed reviewdog suggestions" to collect all the changes that were close connected by topic and time of change, not 4 times "reviewdog" after each other commit for every single change.

It goes for the "restore" commits as well. Why did you put all of them in different commits?

---

> * (1) the suggested max 400 lines per PR would correspond to
>   10456/400 = 26 or more PRs. Calculating with 1 PR per week, that is half year.
>   I hope we can do with much less PRs.

I don't think it makes sense to get too attached to any specific number of lines, as it will depend more on what those changes are.

Reply:
what is the purpose of [DocumentConnection.java](https://github.com/antalk2/jabref/blame/f161808d3f2035ee05e39aa0ea8fbe2a3ba8d225/src/main/java/org/jabref/logic/openoffice/DocumentConnection.java) and why did you create it? Can some/all of it be integrated by itself into the current version of JabRef?


> * (2) a strategy for the presentation might be to build up the alternative
>   version without integrating it, then with the infrastructure in place
>   switch to it. A consequence of this strategy is that until we get to
>   the integration part, the new code is not used, you cannot test
>   it by compiling and running a JabRef. But it is probably easier on the eye.

How does this differ from staggered PRs?

Reply:
> > The last version you can see here is old. Since that I started a new branch, rebased, cut back, then continued.
> > Should I push the current branch?

>     1. You rebased on main?

Yes, two months seems a long time.

There was a change dictated by changes (I think) to some
outside source which the build system "helpfully" upgraded behind my back.
(Aside: this is something I am not used to: code compiles, take a walk, come back,
compilation fails. I am sure it never asked me. Maybe it sent a short message "upgrading x",
I did not read. Since the gradlew build produces random failures anyway, I tried
to recompile a few times before investigating.)
Finally I solved the puzzle: remove a null parameter.
Now I have the official version (from main) at that point.


>        2. Is this your "point A"?

No, with so many moves, it seems that the end result may be
easier to consume if presented as adding new code, integrate, finally
remove what became unused. So PRs will not directly follow the
editing history you see here, you will get a "presentation in packages".

There are some decisions that do have an effect on the "to be presented"
state. Creating a new directory ("logic/oostyle") and associated package
(and corresponding moves under "test") is such a change. I am asking
you to comment on that. Does it seem OK, or should I express the separation
between "style business" and "openoffice business" in a different way?

>     2. With "cut back" you mean that you changed back the code style?

Mostly. I went through the changes against main, and attempted 
to minimize them. Remove those that were not necessary.
Removing a change has the side effect of changing back code style.
I also removed changes to the *.jstyle parser.
Also tried to make the diffs more readable: in OOBibStyle.java I copied back
parts that were removed and just put /* */ around them to help
diff to keep in sync. However this might be confusing for the reader, so
I am not sure if this is the best way. Now I am thinking about
leaving these in initially and use different names for the replacements.

> On the topic of what @calixtus was saying before,
> 
> > "Fixed reviewdog suggestions" to collect all the changes that were close connected by topic and time of change, not 4 times "reviewdog" after each other commit for every single change.
> It goes for the "restore" commits as well. Why did you put all of them in different commits?

I do not claim to have a well established theory and practice of how, when
and what to commit. Cheerful bisecting suggests to commit often, preferably
in compilable/testable states. Reviewing what did I do suggest commit
when something meaningful (nameable) change happened. I usually try
to commit often, but sometimes forget about it.

Restoring different files -- if the changes are local, like formatting or internal changes
not affecting API -- are othogonal to each other. Although I did not do it for those
commits you can see, the changes outside openoffice were "backported" to cancel
out (instead of "change and restore later", we have "never touched" in the history).
I usually did not backport changes inside openoffice, where other changes 
might interact with its threading through history in rebase.

In some cases, like the removed parser changes it might be helpful if later I want
to resurrect some parts. Changes to different files are orthogonal (in diff sense).

Are you asking about the 3 "restore ManageCitationsDialogView.java" commits?
The second follows the first directly could have been augmented instead.
The third follows a bit later, but it seems could be joined to the other two.


> > * (1) the suggested max 400 lines per PR would correspond to
> >   10456/400 = 26 or more PRs. Calculating with 1 PR per week, that is half year.
> >   I hope we can do with much less PRs.
> 
> I don't think it makes sense to get too attached to any specific number of lines, as it will depend more on what those changes are.

I hope so, too.

Reply:
> Yes, two months seems a long time.

Ah, ok, I just wasn't sure what you used rebase for. I find merging easier for keeping up to date with the main branch, even if it doesn't serve exactly the same purpose.

> No, with so many moves, it seems that the end result may be
> easier to consume if presented as adding new code, integrate, finally
> remove what became unused. So PRs will not directly follow the
> editing history you see here, you will get a "presentation in packages".

Perhaps that is better. I think it is what @tobiasdiez suggested as well.

> There are some decisions that do have an effect on the "to be presented"
> state. Creating a new directory ("logic/oostyle") and associated package
> (and corresponding moves under "test") is such a change. I am asking
> you to comment on that. Does it seem OK, or should I express the separation
> between "style business" and "openoffice business" in a different way?

I should have time to take a look at that tomorrow. In practice, I'll try to see if I believe it follows the description in https://devdocs.jabref.org/getting-into-the-code/high-level-documentation.

> DocumentConnection.java collects methods interfacing to UNO API that are not specific
> to JabRef (or OOPanel in JabRef) their only dependences on JabRef
> is (1) thematic: they were collected from/for the OOPanel (2) maybe some exception types
> used.

I see. Thank you for the explanation.

> I presume you mean to move it to somewhere where other parts of JabRef could use it.

I was wondering if that was a good file to make an individual PR from. But if I understand you correctly, that file is intended for personal code-snippets/notes.
I was trying to figure out if there is a way of splitting what you have already done so that you at least can get some concrete comments/recommendations/suggestions on a smaller part of what you have already done. But as mentioned above, perhaps that isn't the right approach.

> Google abandoned me on this. "staggered pull requests" gives only
> https://remysharp.com/2021/03/24/the-mythical-code-review
> which mentions them, but does not explain what they are.
> 
> Is it a synonym for ["stacked pull requests"](https://www.michaelagreiler.com/stacked-pull-requests/)?

Sorry (again), yes, I meant "stacked pull requests".

> I wrote about it in hope of getting some feedback, like "yes, that seems OK",
> or watch out for xxx

Where and on what would you like the feedback?

Reply:
> > No, with so many moves, it seems that the end result may be
> > easier to consume if presented as adding new code, integrate, finally
> > remove what became unused. So PRs will not directly follow the
> > editing history you see here, you will get a "presentation in packages".
> 
> Perhaps that is better. I think it is what @tobiasdiez suggested as well.

Yes, his comment was an important turning point in suggesting
this direction.

> > There are some decisions that do have an effect on the "to be presented"
> > state. Creating a new directory ("logic/oostyle") and associated package
> > (and corresponding moves under "test") is such a change. I am asking
> > you to comment on that. Does it seem OK, or should I express the separation
> > between "style business" and "openoffice business" in a different way?
> 
> I should have time to take a look at that tomorrow. In practice, I'll try to see if I believe it follows the description in https://devdocs.jabref.org/getting-into-the-code/high-level-documentation.

On gui/logic/model:

- It seems, the Openoffice/Libreoffice panel started as a plugin.
- So far, it is a "consumer" in the sense that uses JabRef data types and methods.
  It is not a "provider" in that other parts do not use its methods and data types.
- At some point in history it was split to logic and gui parts, moving code that works together
  and largely independently (not being a provider) far apart. This brougth some uncertainty
  about the status of stuff in logic/openoffice: its location suggests that any part of gui,
(and probably any part of logic?) could depend on it. 

For example, CitationEntry, with its fairly generic name was in logic/openoffice (now in logic/oostyle (? this looks wrong, I'll move it back)).  Its sole purpose is to pass some data from
originally OOBibBase in gui/openoffice, now Backend52->OOFontend->OOBibBase (method getCitationEntries)
to towards ManageCitationsDialogViewModel. It even has its own tests in CitationEntryTest.java. It took some time
to convince myself that it is really not used anywhere else.

What I am trying to suggest: a core vs (plugins and similar leafs) distinction in the hierarchy
might clarify the dependencies. 

> > DocumentConnection.java collects methods interfacing to UNO API that are not specific
> > to JabRef (or OOPanel in JabRef) their only dependences on JabRef
> > is (1) thematic: they were collected from/for the OOPanel (2) maybe some exception types
> > used.
> 
> I see. Thank you for the explanation.
> 
> > I presume you mean to move it to somewhere where other parts of JabRef could use it.
> 
> I was wondering if that was a good file to make an individual PR from. 

I also thought it would be part of one of the first PRs.

> But if I understand you correctly, that file is intended for personal code-snippets/notes.

No, not so personal. There is some code there that is not actually used:
for example printRedlines() and printServiceInfo() are just tools I used and may still
later need to use to investigate, but were never intended to be part of a deployment.
There are some that were used at a point, and may or may not be needed later.
I do not know where should I put these not-for-deployment parts.
Is there, or could I create a place for them somewher inside the tree, or should I move them outside?
The latter means they will not be available for others.

I may be wrong, but I assumed that others share my problems (and confusion) with UNO:
the provided API is raw (reflects the constraints of cross-language communication).

For example what can you do with an XTextCursor, returned by createTextCursor()?
Based on https://www.openoffice.org/api/docs/common/ref/com/sun/star/text/XTextCursor.html
you get 8 methods here, and inherit 6 from XTextRange.

createTextCursor's description mentions that it
 "Returns a new instance of a TextCursor service which can be used to travel in the given text context." 
To use its other capabilities, you cast to its exported interfaces (there are 8 of them)
http://www.openoffice.org/api/docs/common/ref/com/sun/star/text/TextCursor.html
In the end, up to half of the UNO code is calls to UnoRuntime.queryInterface, instead of
getting TextCursor and calling the methods it supports directly.

This cries for a language specific layer. There may be some out there already for java,
I did not check. But it brings in the question of outside dependencies: if there is one,
what should I check before suggesting to include it?

So I assumed that others may find these snippets useful as well. 
The hard part with UNO is often figuring out what is available and how to use it.
For example collecting changes to the document during a GUI action to an "undo context"
is something directly supported by LibreOffice. Easy, if you happen to know
that you need an XUndoManager and how to get one.
3 lines in enterUndoContext, and 3 in leaveUndoContext, of which 2 is identical.
Since nobody in almost a decade installed these calls, it is probably safe to assume
that they are not commonly known. It (the necessity of interface hunting)
is also the reason why I am reluctant to just throw away those that are not currently used.
Someone with a better understanding of UNO and its documentation might feel differently.

Also: someone starting from UNO with the intent to create a language specific
layer would structure differently. What I collected lacks the generality and deeper
understanding of the OO UNO API. For example a general interface could create a class
or interface for TextCursor, and return that from createTextCursor() to make methods directly available.
It is also possible that this would prove a naive approach, bringing too many problems.

> I was trying to figure out if there is a way of splitting what you have already
> done so that you at least can get some concrete comments/recommendations/suggestions
> on a smaller part of what you have already done.

It sounds good to me. The later I get those comments, I'll have more code to rework.
So please, do comment on it. 

> Where

Options? Email?

> and on what would you like the feedback?

This is suprisingly hard to answer. There are some questions I know/feel I need
an answer for (like separating oostyle, where to put not-for-deployment code).
Some of them may be at least partially answered in https://devdocs.jabref.org/ ,
although I may have missed it or did not fully understand.

I did not use java before, so although based on experience with C++ and Haskell I can largely
understand java APIs, the infrastructure and key words MADR, IDEA, Crowdin, slf4j, log4j2,
Gradle, JUnit5, Jackson, DTO,  ... as well as outside references  Java by Comparison,
Effective Java, Design Patterns, and many others pose the question: should I chase this now,
or leave it to later. Often the latter, (otherwise I would still be reading instead of coding).

There is a section [on exceptions and outputting errors](https://devdocs.jabref.org/getting-into-the-code/code-howtos#throwing-and-catching-exceptions)

- All Exceptions we throw should be or extend JabRefException
- Catch and wrap all API exceptions (such as IOExceptions) and rethrow them

UNO throws a bunch of exceptions

- One ot them is [WrappedTargetException](https://www.openoffice.org/api/docs/common/ref/com/sun/star/lang/WrappedTargetException.html) almost any UNO call can throw it, and we don't know what it means "wraps an exception thrown by the original target.". We are just threading it up to the GUI, which usually catches it
and shows to the user. 
- There are others, that may be more informative: PropertyVetoException, which we do nothing about, also threaded
up to GUI.
- There are some we catch and handle, but this is rare.
- and those that should not happen: IllegalArgumentException

How they relate to "Catch and wrap all API exceptions". Where should we do that?
DocumentConnection? OOPanel gui? Somewhere in between? Is threading up everything
in distinct exception types a good thing? A single change at the bottom dictates up to 10 another
to just do the threading. Or should we collect them to some more generic types derived from JabRefException?
Should we do that in DocumentConnection (making it explicitly JabRef-dependent) or in a wrapper around it,
or ...

In one reading all is well, OOPanel catches (tries to catch) everything, thus JabRef proper
is not exposed to these. In another reading it is all wrong.

There are some questions concerning what behaviour  is expected from a user's
point of view, and how my changes relate to other parts of the code.

For example I have been busy implementing pageInfo-for-each-citation. There is a CSL part in the
code, which is also concerned with formatting. Are we trying to connect these two?
Can CSL handle pageInfo-for-each-citation? If we want to offer CSL styles for
OpenOffice, are we going to have two kinds, with different capabilities and/or behaviour?

Migration: can we assume unidirectional (upgrade only), or are we trying to create documents
or work with documents, that older JabRef can deal with? For upgrade-only, we use could use
from pageInfos that are currently in use and throw away those left around. If we have to work with
old-format documents, then assuming "Separate" was already available, the "Merge/Separate" cycle
recovering pageInfos forgotten there, vs "now we do not leave unused data around"
will produce different user-visible behaviuour. Should we keep the forgotten there pagInfos
during an upgrade-then-convert-to-old-style cycle? Upgrade-only is a lot easier to implement. Telling
a user to upgrade their documents seems OK. Telling them to also upgrade all of their coworkers JabRefs
may be too much. Can they use two JabRef versions depending on who they work with?
Stuff in org/jabref/migrations appears unidirectional.

back to
> and on what would you like the feedback?

- There may be (almost certainly, there are) questions I am not aware of, but should know about. Probably concerning
moving or shaping things in a wrong direction.

On the practical side:

I do not have a detailed map of what would go into which PR, not even the first one,
this limits what could you comment on right now.

Things are still moving around. I hope the general direction is discernible.

Reply:
**not-for-deployment parts**

I think I marked private some currently-unused methods in DocumentConnection.
- the good thing about it is that both for persons and for the compiler makes it obvious that it is not used elsewhere.
The compiler can probably eliminate them, seeing the are not used. Unless introspection is required to make private
things available.
- the bad thing is that it does not distinguish private-dont-use-from-outside and just-mark-it-public-if-you-need-it.
This can be confusing without explanation.

**logic/openoffice and gui/openoffice split**

With the split of directories I believe they also became different packages.
I think this latter makes it impossible to express that CitationEntry is private to "openoffice".
Am I missing something here? Modules?  openoffice/logic and openoffice/gui?

Reply:
> **logic/openoffice and gui/openoffice split**
> 
> With the split of directories I believe they also became different packages.
> I think this latter makes it impossible to express that CitationEntry is private to "openoffice".
> Am I missing something here? Modules? openoffice/logic and openoffice/gui?

You are not missing anything. It is not intended to be private to openoffice. You are expressing the class' concern/responsibility (`logic`) and use (`openoffice`) already. Other parts of JabRef can depend on `logic/openoffice` if it is needed in the future.

https://devdocs.jabref.org/getting-into-the-code/high-level-documentation
> The model represents the most important data structures (BibDatases, BibEntries, Events, and related aspects) and has only a little bit of logic attached. The logic is responsible for reading/writing/importing/exporting and manipulating the model, and it is structured often as an API the gui can call and use

Looking at it I am not sure why `CitationEntry` is in `logic/openoffice` instead of `model/openoffice`.

Reply:
> For example, CitationEntry, with its fairly generic name was in logic/openoffice (now in logic/oostyle (? this looks wrong, I'll move it back)). Its sole purpose is to pass some data from

What else would you call it?
It is imported as 
```
import org.jabref.logic.openoffice.CitationEntry;
```
which, in my opinion, makes it quite specific even if the class name is generic.

Reply:
> The hard part with UNO is often figuring out what is available and how to use it.

Yup. Agreed X)
I am currently reading up on UNO and I don't know if there is anyone currently involved with JabRef that has spent time with it. If you have specific questions, there is always the Gitter chat.

Reply:
> But it brings in the question of outside dependencies: if there is one,
> what should I check before suggesting to include it?

https://devdocs.jabref.org/contributing#when-adding-a-library

Reply:
> or should I move them outside?
> The latter means they will not be available for others.

I'd go with this option. As you mentioned, UNO isn't used much in JabRef (that I am aware of) and it will make it easier to review.

Reply:
The reason why CitationEntry is in logic is, is just simply the fact that nobody moved it to a specific model package.
There is simply no `org.jabref.model.openoffice`.
The name itself is fine. 
Regarding UNO:
https://api.libreoffice.org/ (see examples and IDL apidoc)

And https://wiki.openoffice.org/wiki/Documentation/DevGuide/OpenOffice.org_Developers_Guide 
Although it's from OO the general concepts still apply to LO and the usage in JabRef. You can skip the part of the connection to LO because that's a topic on its own which is not relevant.

Reply:
> Options? Email?
> 
> >

An up to date PR but expect the first comments to be about adhering to the codestyle, e.g., [try not to abbreviate names](https://devdocs.jabref.org/getting-into-the-code/code-howtos#generic-code-how-tos). It is to make it easier to review.

Reply:
> like separating oostyle

I don't feel that I understand your goal well enough. Perhaps it will become clearer what class belongs where soon (it is not necessarily always clear-cut).


> How they relate to "Catch and wrap all API exceptions". Where should we do that?
> ...
> Is threading up everything
> in distinct exception types a good thing?

It depends. Many of those exceptions seem to indicate that there is an exception that will prevent what the user is trying to do from happening and therefore should show an error (i.e., it isn't enough with logging it and doing something else).
I'll try to keep an eye out for good examples.

> Or should we collect them to some more generic types derived from JabRefException?

Mostly not, but there are some examples of where it's done, `ShortDOIServiceException`, `FetcherException`, etc.

> In one reading all is well, OOPanel catches (tries to catch) everything, thus JabRef proper
> is not exposed to these. In another reading it is all wrong.

I'd guess that is good enough for most of the exceptions, at least for now.

Reply:
**on CitationEntry**

*Context*

> > You are not missing anything. It is not intended to be private to
> > openoffice. You are expressing the class' concern/responsibility
> > (logic) and use (openoffice) already. Other parts of JabRef can depend
> > on logic/openoffice if it is needed in the future.
> >
> > https://devdocs.jabref.org/getting-into-the-code/high-level-documentation
> >
> >     The model represents the most important data structures
> >     (BibDatases, BibEntries, Events, and related aspects)
> >     and has only a little bit of logic attached.
> >     The logic is responsible for reading/writing/importing/exporting and manipulating the model,
> >     and it is structured often as an API the gui can call and use
> >
> > Looking at it I am not sure why CitationEntry is in logic/openoffice instead of model/openoffice.


>     For example, CitationEntry, with its fairly generic name was in
>     logic/openoffice (now in logic/oostyle (? this looks wrong, I'll move it back)).
>     Its sole purpose is to pass some data from
> 
> What else would you call it?
> It is imported as
> 
> import org.jabref.logic.openoffice.CitationEntry;
> 
> which, in my opinion, makes it quite specific even if the class name is generic.


> The reason why CitationEntry is in logic is, is just simply the fact that nobody moved it to a specific model package.
> There is simply no `org.jabref.model.openoffice`.
> The name itself is fine.

Notes

> It is not intended to be private to openoffice.

> Other parts of JabRef can depend on logic/openoffice if it is needed in the future.

> The model represents the most important data structures

> What else would you call it?

Let's review what is it, and what is it used for?

What is it

```
public class CitationEntry implements Comparable<CitationEntry> {

    private final String refMarkName;
    private final Optional<String> pageInfo;
    private final String context;
    [...] // getters, setters, Comparable<CitationEntry> implementation based on refMarkName
}
```

- `refMarkName` : on return from the gui allows to find the citation group the pageInfo belongs to.
  In the near future it needs to be extended to refer to citations, since pageInfo will be associated
  to citations, not citation groups. A bit further in the future (not part of the current changes)
  reference marks may be replaced with bookmarks, or possibly Comments which can also be listed
  in the LibreOffice Navigator panel, but (I think) do not have names, or, to promote
  interoperability and easy editing they may be replaced 
  with a textual representtaion similar to what JabRefConverter does (also nameless).
  
  To minimize changes I kept refMarkName here, but it basically stands for CitationGroupID.
  To refer to citations CitationPath is used elsewhere.
  
  As long as we need to handle old-style (pageInfo for citationGroups) documents
  in ManageCitationsDialog we need to support both variations.

- `pageInfo` : the only thing that ManageCitationsDialog allows to change

- `context` : a piece of text from the the document: the text of the citation marker
  with up to 30 characters on each side. Changing to pageInfo-for-citations will need
  a change here: "text-before [1-7] text-after" does not tell the user to which of the 7 sources
  does this pageInfo belong to. This also applies to Author-year citations.

Based on this, I would argue that

- CitationEntry is not in any sense among the "the most important data structures"
- I do not see a future where any part of JabRef outside openoffice, or indeed, any part
inside openoffice apart from those implementing
ManageCitationsDialog would want to use this class.
It represents a collection of data for a very particular purpose.

On naming: the only thing that ManageCitationsDialogView allows is:
edit pageInfo records.  It could possibly be extended with a
"link-to-location-in-the-document", and sorting (by cited source / by
document order). To do more (what actually?) would probably require
further dialogs, or at least a major revision of the current GUI.

I am not even sure that ManageCitationsDialog (or, to reflect what
it does, "PageInfoEditor") is the right direction. Both Jabref and
LibreOffice GUIs are built with a "the whole screen is mine" mindset,
(even the 2nd demo gif  [here](https://docs.jabref.org/cite/openofficeintegration)
struggles showing them together).
It is probably best if we try to minimize the number of required
switches between them. Although ManageCitationsDialog is a dialog with its
own window, so it might fit on the same screen as LibreOffice,
its "parallel" use with LibreOffice is limited,
because it shows a snapshot of the document: if the user edits the text (he
cannot do much, but may remove some citation groups), the state in
ManageCitationsDialogView becomes outdated (it is unavoidable, unless
we can hook into LibreOffice to "let me know when this reference mark
is deleted", or start polling). If textual representation (similar to
JabRefConverter's, but extended to include pageInfo) comes around, the
PageInfoEditor functionality may probably better served by that: it
would make moving or copying citations between locations as well as
editing the pageInfo parts a more natural experience for
users of LibreOffice, as opposed to opening extra dialogs from an external
program (JabRef) for these purposes.

Clickable links to the list of citation groups (reference marks) are already provided
by the LibreOffice Navigator (we could add bookmarks for the discovered textual parts, too).
We could add links from the bibliography entries
to the citation groups containing citations of a given source (an exploratory version
of this is in the code, only shown for Unresolved sources to stay close to the earlier
output in the normal case).

**Summary**

- CitationEntry is unlikely to be of interest outside its current use
(passing around a collection of data collected for ManageCitationsDialog)

- Depending on the future of ManageCitationsDialog
it may need modifications to serve the extended purpose or may become
unused if in-LibreOffice solutions make ManageCitationsDialog unnecessary.

- Naming CitationEntry: something like ManageCitationsDialogEntryData.
  (there is also "CitationEntryViewModel.java" in gui/openoffice).

**separating oostyle**

> > like separating oostyle
> 
> I don't feel that I understand your goal well enough.
> Perhaps it will become clearer what class belongs where soon (it is not necessarily always clear-cut).

The direct goal is separation of concerns.

*oostyle*: given a .jstyle, a list of citation groups (in order of
  appearance) and a list of citations (citation keys) in each group,
  and access to some BibDatases, produce citation
  markers for these groups and a bibliography according to the style. It needs to communicate
  formatting instructions, uses OOFormattedText for this purpose.

  In principle this functionality does not depend on if we use LibreOffice,
  OpenOffice, MS Word, or generate a HTML page, as long as we can feed it the necessary data,
  and interpret the markup it produces into the target.

  It might also produce cross-references between citations and bibliography entries
  by including links in the markup.

  I expect CSL does something like this (style+citations and cites to markup) as well.

*logic/openoffice* contains a mixture of:

- (1: backend) how do we store/retrieve locations of
  the citation groups (reference marks) and list of citations
  belonging to them (reference mark names) and associated data
  (citep,citet,nocite) in reference mark names, pageInfo in
  user-defined document properties), as well as how do we find the
  taxt range containing the bibliography for refresh.
- (2: OOFormattedText interpreter) the markup received from oostyle needs to be interpreted
  into the locations the backend points to.
- (3) methods to be called from GUI. Those that change the document need to communicate to the backend.

So oostyle is a part that might be of interest in other contexts.
Correspondingly its main data structures could probably move to "model/oostyle".

From logic/openoffice it is harder to pick.

`DocumentConnection` is not really a data structure. It is basically an
XTextDocument with some of its relations and interfaces cached in its
other fields. Just as we do it in the constructor, we could always
get them starting from the XTextDocument.  The rest is mainly a
collection of methods, some of them already static, others use the
XTextDocument and/or the cached values, but could be made static by
adding an XTextDocument argument. And it has a LOGGER.

`OOUtil` is mostly concerned with interpreting `OOFormattedText` into
an OpenOffice or LibreOffice writer document (`insertOOFormattedTextAtCurrentLocation2`)

It has a small counterpart within oostyle/OOFormat.java : `setLocale`, `setCharStyle`, and `paragraph`
wrap an `OOFormattedText` into the corresponding tags. Most of the markup comes
from the layout and other options from the jstyle.

The rest of logic/openoffice seems more or less specific to either the backend or the gui,
with OOFrontend in between. I expect to move more support-the-gui code
from OOBibBase here. RangeSortVisual provides the "order of appearance" of
citation groups needed by oostyle. RangeKeyedMapList might possibly be replaced
by or reimplemented using a MultiMap from elsewhere.

**abbreviations in names**

> e.g., [try not to abbreviate names](https://devdocs.jabref.org/getting-into-the-code/code-howtos#generic-code-how-tos).

It says: "try not to abbreviate names of variables, classes or methods"

- The general principle, as I saw before was closer to: try not to abbreviate **global** names.
- The hard part is: what is the expected balance between avoiding unnecessary verbosity
that just clutters the code and cryptic (or well-known?) abbreviations 
(like awt, fxml, slf4j, cli, gui, i (in StringInt),
CSLType, RisMonth, AuthorLF_FF, JabRef :). Unfortunately "cryptic"
depends on who is seeing it, with what background and which state of mind
(as in looking at ones' own code a year later).
- Below [try not to abbreviate names](https://devdocs.jabref.org/getting-into-the-code/code-howtos#generic-code-how-tos)
I find `catch (IOException ioe)` in an official example. This suggests
that local variables are probably exempt.
`IOException` itself contains an abbreviation, and is a globally used name.

My filenames (and corresponding classnames)  are usually
not abbreviations (except the "OO" prefix already in use before,
`Compat` (would it be more expressive as `Compatibility`?)
How does it relate to contractions in
"the most important data structures (`BibDatabases`, `BibEntries`, ...)",
`getPrefsNodeForCustomizedEntryTypes`, `ParamLayoutFormatter`?

- Seeing `for (LayoutFormatter anOption : option) {` reveals that the name `option`
refers to a collection of things with type `LayoutFormatter`.
These are named `anOption` which contributes nothing to my understanding. Nor would it
help if we called it `aLayoutFormatter` or `theCurrentLayoutFormatter`

- My local variables and argument names often are abbreviated: for example `CitationGroup cg`.
Replacing `cg` with `citationGroup` is about as useful as
changing `IOException ioe` to `IOException ioException`:
adds nothing to what we already know from the verbose typename.

I will probably need more explicit examples of which abbreviations
should I expand. I, of course understand them (for now), but it is hard to guess
which ones could appear confusing or cryptic to others, or even myself at another time.

Reply:
> I will probably need more explicit examples of which abbreviations
> should I expand. I, of course understand them (for now), but it is hard to guess
> which ones could appear confusing or cryptic to others, or even myself at another time.

Not all JabRef code is up to date, there are unfortunately still parts of the code where the standard is lower than we'd like it to be. When it comes to `AuthorLF_FF` and some other artifacts of `Authors.java`/`BracketedExpression`, they are a mix of code that should be refactored, but no one has had the time to do so, and names/commands that the user is expected to manually enter within JabRef or in an external text editor, without autocompletion, type safety, etc. They probably would have been named differently if they were created today.

 Imho `https://github.com/antalk2/jabref/blob/improve-reversibility-rebased-03/src/main/java/org/jabref/logic/openoffice/Backend52.java#L236-L240`
```
CitationGroup cg = new CitationGroup(cgid,
                                                 sr,
                                                 itcType,
                                                 citations,
                                                 pageInfo);
```
and similar would be helpful (to me) if they were expanded.

Generally speaking, I'd make a separation between what I'd consider ok in a draft-PR and what I'd consider ready to merge, and I'd still try to err on the side of expanding names, as opposed to abbreviating them. I can't say that I have never written `IOException e` but I probably should at least have done `IOException exception` 😬